warehouse/upload.go

```diff
@@ -683,142 +700,39 @@ func (job *UploadJobT) exportRegularTables(specialTables []string, loadFilesTabl
-		}
-	}
-	return res
-}
-
-// TableUploadStatusT captures the status of each table upload along with its parent upload_job's info like destination_id and namespace
-type TableUploadStatusT struct {
-	uploadID      int64
-	destinationID string
-	namespace     string
-	tableName     string
-	status        string
-	error         string
-}
-
-// TableUploadStatusInfoT captures the status and error for [uploadID][tableName]
-type TableUploadStatusInfoT struct {
-	status string
-	error  string
-}
-
-// TableUploadIDInfoT captures the uploadID and error for [uploadID][tableName]
-type TableUploadIDInfoT struct {
-	uploadID int64
-	error    string
-}
+func (job *UploadJobT) TablesToSkip() (map[string]model.PendingTableUpload, map[string]model.PendingTableUpload, error) {
+	job.pendingTableUploadsOnce.Do(func() {
+		job.pendingTableUploads, job.pendingTableUploadsError = job.pendingTableUploadsRepo.PendingTableUploads(
+			context.TODO(),
+			job.upload.Namespace,
+			job.upload.ID,
+			job.upload.DestinationID,
+		)
+	})
 
-func (job *UploadJobT) fetchPendingUploadTableStatus() []*TableUploadStatusT {
-	if job.tableUploadStatuses != nil {
-		return job.tableUploadStatuses
+	if job.pendingTableUploadsError != nil {
+		return nil, nil, fmt.Errorf("pending table uploads: %w", job.pendingTableUploadsError)
 	}
-	sqlStatement := fmt.Sprintf(`
-		SELECT
-		  UT.id,
-		  UT.destination_id,
-		  UT.namespace,
-		  TU.table_name,
-		  TU.status,
-		  TU.error
-		FROM
-		  %[1]s UT
-		  INNER JOIN %[2]s TU ON UT.id = TU.wh_upload_id
-		WHERE
-		  UT.id <= '%[3]d'
-		  AND UT.destination_id = '%[4]s'
-		  AND UT.namespace = '%[5]s'
-		  AND UT.status != '%[6]s'
-		  AND UT.status != '%[7]s'
-		  AND TU.table_name in (
-			SELECT
-			  table_name
-			FROM
-			  %[2]s TU1
-			WHERE
-			  TU1.wh_upload_id = '%[3]d'
-		  )
-		ORDER BY
-		  UT.id ASC;
-`,
-		warehouseutils.WarehouseUploadsTable,
-		warehouseutils.WarehouseTableUploadsTable,
-		job.upload.ID,
-		job.upload.DestinationID,
-		job.upload.Namespace,
-		model.ExportedData,
-		model.Aborted,
-	)
-	rows, err := job.dbHandle.Query(sqlStatement)
-	if err != nil && err != sql.ErrNoRows {
-		panic(err)
-	}
-	defer func() { _ = rows.Close() }()
 
-	tableUploadStatuses := make([]*TableUploadStatusT, 0)
+	var (
+		previouslyFailedTableMap   = make(map[string]model.PendingTableUpload)
+		currentlySucceededTableMap = make(map[string]model.PendingTableUpload)
 
-	for rows.Next() {
-		var tableUploadStatus TableUploadStatusT
-		err := rows.Scan(
-			&tableUploadStatus.uploadID,
-			&tableUploadStatus.destinationID,
-			&tableUploadStatus.namespace,
-			&tableUploadStatus.tableName,
-			&tableUploadStatus.status,
-			&tableUploadStatus.error,
-		)
-		if err != nil {
-			panic(err)
+		// Previous upload and table upload failed
+		exportingfailedStatus = func(status string) bool {
+			return status == TableUploadExportingFailed || status == UserTableUploadExportingFailed || status == IdentityTableUploadExportingFailed
```

Comment:
[minor] & opinionated
```suggestion
                         switch status {
                         case TableUploadExportingFailed, UserTableUploadExportingFailed, IdentityTableUploadExportingFailed:
                             return true
                          default:
                             return false
                          } 
```

Reply:
We should consider moving this logic outside of this function. Maybe it could be part of the table upload model, or introduce a separate failure flag. 

The concern I have is when we introduce a new failed state we might forget to add it here.

Reply:
Good point.

Reply:
Let me make the change.

Reply:
Looks like the other 2 flags were not being used. Removing this.